README.md
# niMan  
### aka NIM (Node Installation Manager)
[![npm version](https://badge.fury.io/js/niman.svg)](https://badge.fury.io/js/niman)    
[![NPM](https://nodei.co/npm/niman.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/niman/)  

WARNING:

  BETA release so don't expect it to work bugless ..
  
  and is it still just intented to be used on Windows ... 
  (as replacement for NVM)

    although it is already working,  
     it still doesn't override your main version of node, that you have already installed  
      (which lets this one was able to boot),  
      In order to have NIM take over, 
       you have to put the 'inuse' folder (which is inside the installation folder),
        inside the PATH environment variable, and make sure it comes before your original node installation path, 
         finally, if you installed it with NPM you may whant to use 'nim root' command,  
          to know where actually is such folder .. 
:D

# Installation - Usage


after cloning the repository wherever you'd like to ..  
(or using NPM [''npm install -g niman''])  

just type:  
```
nim
```  
at the command prompt ...
  this will either install required node and packages  
  or just start the app showing the Usage information ...  
  (actually you can use whichever command argument you'd like to ..)
  
NIM uses a two stages loader, so it first checks for his own version of Node (6.2.1, at the moment )  
and if not present, downloads it along with corresponding NPM ...   
then installs additional ´required´packages,   
and finally starts the app as would do it normally ...  
  
![](demo.gif)


# GUI  

![](GUI.png)


CHANGELOG.md
## Change Log


### v0.0.25 (2017/03/21)
- just added some documentation ..  
  (like this one)

### v0.0.24 (2017/03/21)
- breakthrough version which now works without any third party dependency ...  
  basically we removed request and express dependency  
  (they used nearly 200 packages, nearly 900 files and an huge space of almost 5MB!)  
  
  


GUI.js
 const
            info            =   {
                                version :   '0.0.1'
                            ,   desc    :   'Node Installation Manager GUI'
                            }
        ,   express         =   require ('express')
        ,   morgan          =   require ('morgan')
        ,   socketIO        =   require ('socket.io')
        ,   _log            =   (...args)       =>  { Function.apply.call(console.log   ,console,args); }
        ,   _err            =   (...args)       =>  { Function.apply.call(console.error ,console,args); }
        ;                    
    //  ===================================================== GUI Configuration
    const   _APP                =   {};
            _APP.PUBLIC_HTML    = '/GUI'; 
            _APP.BOWER_DIR      = '/js-libs';
            _APP.LISTEN_PORT    = process.env.PORT || 1111;
            _APP.URL_BASE       = 'http://localhost';
            _APP.log            = _log
            _APP.err            = _err
            _APP.timeSt         = (name)  => { return timers[name]= (new Date()).getTime();};
            _APP.timeEn         = (name)  => { return (new Date()).valueOf() - timers[name];};
            _                   = _APP;
   
    //  ===================================================== 
    let     ND
    ,       me              =   process
    ,       show            =   true
    ;
    me.on('exit', function (c) { process.exit(c); })


             try{   
                app=express();
                // setup static content folders 
                app.use(express.static(__dirname + _.PUBLIC_HTML));                 //
                app.use(express.static(__dirname + _.BOWER_DIR  ));
    
   
                app.use('/cmmnd', (reQ,res) => {// _log('reQ',reQ);
                    var H   = reQ.headers
                    ,   R   = reQ.rawHeaders
                    ,   U   = reQ._parsedUrl
                    ,   fn  = U.query
                    ;
                    switch (fn) {
                      case 'arch' :
                                    var arch = do_ARCH(fn);
                                    if (arch) res.send(arch);
                                    else     res.send('ERROR!');
                                    break;
                      case 'list' :
                                    var list = do_LIST(fn);
                                    if (list) res.send(JSON.stringify(list));
                                    else      res.send('ERROR!');
                                    break;

                        default:
                            res.send('unknown command');
                    }
                    
                });    
                
                morgan && app.use(morgan('dev'));                                   // log every request to the console
                app.use(function(req, res){ res.sendStatus(404);});                 // simply NOT FOUND
                // ====================================================== Main 
                server = app.listen(_.LISTEN_PORT,  () => {
                  _.log('Express server listening on '+_.URL_BASE+':'+_.LISTEN_PORT+'/');
                }); 

                
                io     = socketIO.listen(server);
                
                if (!show){
                 var child=childProc.spawn("cmd",['/C','START',_.URL_BASE+':'+_.LISTEN_PORT, cmd] );
                    child.on('error',function (err) { _err(err);    me.exit(-123);  });
                    child.on('exit', function (code){ _log(); });
                }
             }
             catch (ex) {
                 _err(ex);
                 me.exit(-123);
             }       
                            

GUI/js/main.js
    var ND 
    ,   _               =   _ || {}
    ,   leftPad         =   _padLeft.func
    ,   _W              =   window        
    ,   _D              =   document
    ,   _O              =   Object
    ,   _C              =   console
    ,   _B              =   _D.body
    ,   _FN             =   Function.apply
    ,   _ById           =   function  (id)                      { return _D.getElementById (id);        }
    ,   _ByQs           =   function  (id)                      { return _D.querySelector  (id);        }
    ,   _newHtmlEl      =   function  (el)                      { return _D.createElementNS  ('http://www.w3.org/1999/xhtml', el);   }
    ,   _newSvgEl       =   function  (el)                      { return _D.createElementNS  ('http://www.w3.org/2000/svg'  , el);   }
    ,   _newEL          =   function  (elm,id)                  { var el=_D.createElement    (elm); if (id) el.id=id; return el;     }
    ,   _newTXT         =   function  (txt)                     { var el=_D.createTextNode   (txt); return el;                       }
    ,   _setAttr        =   function  (e,a,v)                   { e.setAttribute(a,v); return e;        }
    ,   _Ajax           =   function  (req , cb , err, sts )    {
        var     xhttp;
        
        if (window.XMLHttpRequest) {
            xhttp = new XMLHttpRequest();
            xhttp.timeout = 12000;
        } 
        else {
            // code for IE6, IE5
            xhttp = new ActiveXObject("Microsoft.XMLHTTP");
        }

        
        xhttp.onreadystatechange = function() {
            if (sts) sts(xhttp);
            switch (xhttp.readyState ) {
            
             case 4: {
              switch(xhttp.status) {
                case 200: 
                        cb    (xhttp.responseText); 
                        break;
                 default: 
                        err   ({'sts' : xhttp.status ,'txt' : xhttp.responseText}); 
                        break;
              }
             }
             default:
               //log(xhttp.readyState,xhttp.status);
            }   
        };
       
        xhttp.open("GET", req, true);
        
        try {
         xhttp.send();
        }
        catch (e) { log (e); }    
    }
    ,   getLang         =   function  ()                        {

        if (navigator.languages !== ND) return navigator.languages[0]; 

        else return navigator.language;

    }
    ,   toLocale        =   function  (ds)                      { 

            var da      = ds.split("-")

            ,   date    = new Date(da[0], da[1]-1,da[2])

            ;

            return date.toLocaleDateString(getLang(),{ 'month' : 'short','year':'numeric','day':'numeric'})

    }    
    ,   _newHtmlEl      =   function  (el)                      { return _D.createElementNS  ('http://www.w3.org/1999/xhtml', el);   }
    ,   _log            =   function  ()                        { _FN.call(console.log, console, arguments); }
    ,   _OrEmpty        =   function  (v)                       { return v?v:''; } 
    ,   showError       =   function  (lbl, err)                {
            var tbl=_ById('list');
                  
                  tbl.innerHTML+='<TR><TD colspan=10>'+lbl+'ERROR: ' + err.sts+'&nbsp;'+ (err.txt?err.txt:'') +'</TD></TR>';
    }
    ,   onlyMainVersions=   function    (list)                  {
        var vi = {}
        ,   vl = []
        ,   v
        ,   i
        ,   hiv
        ,   n
        ;
        for (i in list){
            v = list[i];
            n = numeric(v.version);
            hiv = n.substring(0,3);
            vo  = vi[hiv];
            if (vo==ND || (vo.version - v.version)>0)   vi[hiv]=v;  
            
        }
        for (i in vi) vl.push(vi[i]);
        return vl;
    }
    ,   showList        =   function    (list)                  {
         list=onlyMainVersions(list);
         
            var tbl=_ById('list');
                 for (v in list){
                  var ver=list[v];
                  tbl.innerHTML+='<TR><TD><A target=dist href="' + (ver.origin == 'NODE' ? 'https://nodejs.org/dist/' : 'https://iojs.org/dist/') + ver.version + '/">' +ver.version+'</A>'
                                +'</TD><TD>'+_OrEmpty(ver.origin            )
                                +'</TD><TD>'+_OrEmpty(toLocale(ver.date)    )
                                +'</TD><TD>'+_OrEmpty(ver.lts               )
                                +'</TD><TD>'+_OrEmpty(ver.npm               )
                                +'</TD><TD>'+_OrEmpty(ver.v8                )
                                +'</TD><TD>'+_OrEmpty(ver.openssl           )
                                +'</TD><TD>'+_OrEmpty(ver.zlib              )
                                +'</TD><TD>'+_OrEmpty(ver.uv                )
                                +'</TD></TR>';
                 }
    }
    ,   getNodeListInProgress   = true
    ,   getIoJsListInProgress   = true
    ,   nodeList
    ,   nodeError
    ,   iojsList
    ,   iojsError
    ,   numeric         =   function    ( vs )                  {
        vs = vs.substring(1);
        vs = vs.split('.');
        var len = vs.length;
        if (len!=3) return 0;
        var num= leftPad(vs[0],3,0) + leftPad(vs[1],3,0) + leftPad(vs[2],4,0) ;
        return num;
        
    }
    ,   compareVersion  =   function    ( nV ,iV)               { return  numeric(nV) > numeric(iV); }
    ,   raimbowCol      =   function    (i, ofMax, r, ph)       {
     var    
            f   = 360 / ofMax
     ,      rd  = Math.PI / 180
     ,      mid = 128
     ,      ph  = (ph ||  40) % 360
     ,      r   = (r  || 127) 
     ,      R   = Math.round(Math.sin((f*i-  0+ph)*rd) * r + mid)
     ,      G   = Math.round(Math.sin((f*i-120+ph)*rd) * r + mid)
     ,      B   = Math.round(Math.sin((f*i-240+ph)*rd) * r + mid)
     ,      c   
     ;
      R= R>160?255:R>96?127:0;
      G= G>160?255:G>96?127:0;
      B= B>160?255:B>96?127:0;
      c = (R*0x10000+G*0x100+B)
      return c;
    }
    ,   addYearMark     =   function    (dv, date,start,step)   {
        var mrk = _newHtmlEl('div')
            mrk.setAttribute('class','yearMark');
            
        var x= (date.getTime()-start) * step;
            mrk.innerHTML=date.getFullYear();
            mrk.setAttribute('style','left: '+(x>>0)+'px; top:0px; height:100%;');
            dv.appendChild(mrk);
    }
    ,   showGraph       =   function    (values,col)            {
        var my      = this
        ,   dv      = _ById('svgDiv')
        ,   bound   = dv.getBoundingClientRect()
        ,   W       = bound.width
        ,   H       = bound.height
        ,   len     = values.length
        ,   min     = new Date(values[len-1].date).getTime()
        ,   max     = new Date(values[0]    .date).getTime()
        ,   step    = (W-64)/(max-min)
        ,   i       = 0
        ,   x       = 0
        ,   y       = 0
        ;
        
        addYearMark(dv,new Date('1-1-2012'),min,step);
        addYearMark(dv,new Date('1-1-2013'),min,step);
        addYearMark(dv,new Date('1-1-2014'),min,step);
        addYearMark(dv,new Date('1-1-2015'),min,step);
        addYearMark(dv,new Date('1-1-2016'),min,step);
        addYearMark(dv,new Date('1-1-2017'),min,step);
        
        var alt     = _newHtmlEl('div')
        ,   altTxt  = null
        ;
            alt.setAttribute('class','altLabl');
            document.body.appendChild(alt);
        
        for (i=0; i<len; i++) {
         var val =   values[len-i-1];
         y = H-numeric(val.version)   / 100000000*H *1.01
         x = (new Date(val.date).getTime()-min) * step
         
         
         var    col     = raimbowCol(i,len/2).toString(16)
                col     = '0'.repeat(6-col.length)+col;
         var    el     = _newHtmlEl('div')        
                el.setAttribute('class','verLab');
                el.setAttribute('style','left: '+((x>>0))+'px; top:'+((y>>0)-32)+'px; color:#'+col+';');
                if (val.origin=='NODE')
                      el.innerHTML='<a target=DOCs href="https://nodejs.org/docs/'+val.version+'/api/">'+val.version+'</a>';
                else  el.innerHTML='<a target=DOCs href="https://iojs.org/docs/'  +val.version+'/api/">'+val.version+'</a>';
                
                el.addEventListener('mouseenter', function (e){
                   if (!altTxt) { 
                    alt.innerHTML= altTxt = this.innerHTML; 
                    alt.style.visibility = 'visible';
                   }
                   this.xx=e.clientX;
                   this.yy=e.clientY;
                });
                
                el.addEventListener('mousemove', function (e){
                   alt.style.top  = ''+ (e.clientY -48)  +'px';
                   alt.style.left = ''+ (e.clientX -48)   +'px';
                   
                });
                
                el.addEventListener('mouseout', function (e){
                   alt.style.visibility = 'hidden';
                   altTxt = null;
                   alt.innerHTML = '';
                });
                
                dv.appendChild(el);
        } 
        
        
    }
    ,   onAcquiredLists =   function    ()                      {
        
            if (getNodeListInProgress || getIoJsListInProgress) {
                if (nodeError) {}
                if (iojsError) {}
                return; // not having both ready!
            }    
            
                var total   =   0
                ,   nodeLen =   nodeList? nodeList.length : 0
                ,   iojsLen =   iojsList? iojsList.length : 0
                ,   i
                ,   j       =   0
                ,   k       =   0
                ,   merged  =   []
                ,   empty   =   'v'
                ,   smmry   =   _ById('smmry');
                ;
                
                
                if (nodeError) { log(nodeError); showError('NODE:',nodeError); }
                else total += nodeLen;
                
                if (iojsError) { log(iojsError); showError('IOJS:',iojsError); }
                else total += iojsLen;
                
                for (i=0; i < total; i++) {
                  var nodeVer = (nodeList && nodeList[j]) ?  nodeList[j].version : empty;
                  var iojsVer = (iojsList && iojsList[k]) ?  iojsList[k].version : empty;
                 
                  if (compareVersion(nodeVer,iojsVer))  { merged[i]=nodeList[j++]; merged[i].origin='NODE'; }
                  else                                  { merged[i]=iojsList[k++]; merged[i].origin='IO.JS'; }
                }
                _log (j,k, merged.length);
                
                smmry.innerHTML = "Found <b>"+merged.length+"</b> releases ..<br>  ("+j+") from NODE and ("+k+") from IO.JS"
                                ;
                
                
                showList(merged);
                
                showGraph(merged);
    }
    ,   doUse           =   function    (el)                    {
            alert('doUse: '+el );
    }
    ;
    
    _onDocReady ( function () { _log('DOM is Ready ..');

    
            _Ajax ('/cmmnd?list' 
            , function (res) { 
                var list        =   JSON.parse(res)
                ,   inUse       =   _ById('inUse')
                ;
                _log('installed list',list);
                
                for (i in list){
                  var oo=list[i];
                  for (v in oo) if (v!='xyw') break;                  
                  inUse.innerHTML   +=  '<TR '+(oo.xyw?'class=inUseV':'') +'>'
                                    +    '<TD><div onclick=doUse(this) class="     btn paper paper-lift">'+v             +'</div></TD>'
                                    +    '<TD><div                     class="arch btn paper paper-lift">'+oo[v]    +'</div></TD>'
                                    +  '</TR>';
                }
                    
            }
            , function (err) { _log('ERR:',err); }
            );
            
    
            _Ajax ('node_dist.json'  //'http://nodejs.org/dist/index.json' 
            , function (res) { 
                var list        =   JSON.parse(res);
                    nodeList    =   list;
                    getNodeListInProgress=false;
                    _log('got Node list');
                    onAcquiredLists();
            }
            , function (err) { nodeError=err; _log('ERR:',err); onAcquiredLists(); }
            );
            
            _Ajax ('iojs_dist.json'  //'https://iojs.org/dist/index.json'
            , function (res) { 
                var list        =   JSON.parse(res);
                    iojsList    =   list;
                    getIoJsListInProgress=false;
                    _log('got IoJs list');
                    onAcquiredLists();
            }
            , function (err) { iojsError=err; _log('ERR:',err); onAcquiredLists(); }
            );
            
    });  
            


js-libs/string_prototypes/index.js
'use strict';


var _String_Prototypes  =   {
                                Name:       'String Prototypes (aka es6-strings)' 
                            ,   Desc:       'Provides ES5 and ES6 compliant prototype functions, along with helpful others ..'
                            ,   Version:    '0.0.3'
                        }
                        // 'str.'.repeat        (3)         ->      'str.str.str.'
                        // 'str.'.startsWith    ('st')      ->      true
                        // 'str.'.endsWith      ('r.')      ->      true
                        // 'str.'.includes      ('tr')      ->      true
                        
                        // 'file.com'.subStr    (-4)        ->      '.com'
                        // 'file.com'.subStr    (-3,-1)     ->      '.'
                        // 'file.com'.subStr    (-4, 1)     ->      '.'
                        // 'file.com'.subStr    ( 4,-4)     ->      'file'

_String_Prototypes.apply = function (scope) {
    var ND // undefined
    ,   th  = scope.String.prototype
    ,   sbs = 'substring'
    ,   ln  = 'length'
    ;
    
    if (!   th.subStr       )   { // handy extension to substring funcionality  subStr(-3) subStr(-3,2)
            th.subStr  =    function subStr (st,len) {
                var s = st  == ND ? 0         : st  < 0 ? this[ln]+st : st;
                var e = len == ND ? this[ln]  : len < 0 ? s+len       : st < 0 ? s+len : len;
                return this[sbs](s,e);
            }    
    }
    
    if (!   th.repeat       )   { // ES6 ! 
            th.repeat   =   function repeat (n) {
            var res =   ''
            ,   me  =   this
            ;
            
            do {
                res += (n & 1) ? me : '';
                me  += me;
                n>>>=1;    
            }
            while (n);
            return res;
        }
    }         
        // we silently convert s to string ... flexibility gives always a better user experience
    
    if (!   th.startsWith   )   {   // ES6 !
            th.startsWith   =   function    startsWith  (s) { 
                var re = new RegExp('^'+s);
                return re.test(this);
            }     
    }
    
    if (!   th.endsWith     )   {   // ES6 !
            th.endsWith     =   function    endsWith    (s) {
                var re = new RegExp(''+s+'$');
                return re.test(this);
            }     
    }

    if (!   th.includes     )   {   // ES6 !
            th.includes     =   function    include     (s) {
                var re = new RegExp(''+s);
                return re.test(this);
            }     
    }

    if (!   th.codePointAt  )   {   // ES6 !
            th.codePointAt  =   function    codePointAt (p) {
                var me  = this
                ,   len = me.len
                ,   p   = p ? Number(p) : -1
                ;
                if (    p != p //NaN
                    ||  p  < 0
                    ||  p >  len
                    )               return undefined;
                var fst = me.charCodeAt(p)
                ,   sec 
                ;    
                if ( fst >= 0xD800 && fst <= 0xDFFF && len > (p+1)) {
                    sec = me.charCodeAt(p+1);
                    return      ((fst - 0xD800) <<  11)
                            +   ((sec - 0xDC00) +   0x10000)
                }
                return fst;
            }     
    }

    if (!   th.trim         )   {   // ES5 !
            (function() {
                var rtrim = /^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g;
                String.prototype.trim = function() {
                    return this.replace(rtrim, '');
                };
            })();
    }; 
    
 return this;    
};

if (typeof module !== 'undefined' && module.exports)  module.exports  = _String_Prototypes;



bootLoader.js
'use strict';

// Hi, I'm the niMan Boot Loader ...
//
//  ------------------------------------------------------------------------------------------------
    var ND                                                                                              // undefined!
    ,   me              =   process
    ,   os              =   require ('os')
    ,   fs              =   require ('fs')  
    ,   http            =   require ('https')
    ,   childProc       =   require ('child_process')
    ,   exec            =   childProc.execFileSync 
    ,   w4it            =   require ('./js-libs/w4it')
    ,   isOS            =   me.arch
    ,   args            =   me.argv
    ,   launchPrfx      
    ,   launchSffx      
    ,   bootFile        =   'nim'
    ,   myNodeFile      =   'node'
    ,   myNodeFile2                                                                                     //  windows only
    ,   myNodePath      =   'my'
    ,   myNodeFileExt                                                                                   //  windows only ?
    ,   myNodeVer       =   'v6.10.0'                                                                   //  LTS Boron
    ,   myNPM           =   'v3.10.10'
    ,   npmExtractPath
    ,   npmPath
    ,   myNpnZipFile   
    ,   npmCommandFile
    ,   npmTempPath
    ,   nodeDistBase    =   'https://nodejs.org/dist/'
    ,   npmDistBase     =   'https://codeload.github.com/npm/npm/zip/'
    ,   nodeDistURI 
    ,   sts1
    ,   sts2
    ,   sts3
    ,   distFolder
    ,   weNeedDownload
    ,   fileToDown
    ,   dwnInProg1      =   true
    ,   dwnInProg2      =   true
    ,   dwnInProg3      =   true
    ,   _log            =   function    ()                  {   Function.apply.call(console.log     ,console,arguments); }
    ,   _err            =   function    ()                  {   Function.apply.call(console.error   ,console,arguments); }
    ,   exit            =   function    (msg,err)           {
        var extc    = msg ? -234 : 0;
        extc && _err(msg,err);
        me.exitCode = extc;
        me.exit();
    }
    ,   linkDirSync     =   function    (srcDir,trgDir)     {
        if (!fs.existsSync(trgDir)){
            exec('cmd',['/C','mkDir',toLocalOs(trgDir)],{  stdio: 'inherit' } );
        }
        exec("sudo.cmd",["mklink","/D",toLocalOs(trgDir+'/'+myARCH),toLocalOs(__dirname+'/'+srcDir)],{  stdio: 'inherit' } );
    }
    ,   copyDirSync     =   function    (srcDir,trgDir)     {
        exec("cmd",["/C","xcopy","/E","/J","/I",toLocalOs(srcDir),toLocalOs(trgDir)],{  stdio: 'inherit' } );
    }
    ,   copyFileSync    =   function    (srcFile,trgFile)   {
        exec("cmd",["/C","copy",toLocalOs(srcFile),toLocalOs(trgFile)],{  stdio: 'inherit' } );
    }
    ,   toLocalOs       =   function    (s)                 {
        switch (isOS){
            case    'x86':
            case    'x64':
                return s.replace(/\//gi,'\\');
            default:
                return s;
        }
    }
    ,   download        =   function    (url, dest, cb)     {   _log('downloading: ',url);
    
        var file    = fs.createWriteStream(dest)
        ,   request = http.get(url, function(response) {
            var sts=response.statusCode;
            if ( sts >= 300 && sts < 400){
                _log('got a redirection: ',response.headers)
            }
            if (sts !== 200) {
                fs.unlinkSync(dest); 
                return cb('Response status was ' + sts);
            }
            response.pipe(file);
            file.on('finish', function() {  file.close(cb); });
        });

        request.on('error', function (err) {    
            fs.unlinkSync(dest);
            if (cb) return cb(err);
        });

        file.on('error', function(err) { 
            fs.unlinkSync(dest); 
            if (cb) return cb(err);
        }); 
    }
    ,   myARCH          =   'win-'+isOS;
    ;
    
    let __CWD=process.cwd();
    me.chdir(__dirname);            // work inside local folder!
    
    switch (isOS) {
        case    'x86':
        case    'x64':
                        nodeDistURI     =   nodeDistBase+myNodeVer+'/'+myARCH+'/';
                        myNodeFile2     =   myNodeFile+'.lib';
                        myNodeFile     +=   '.exe';
                        myNpnZipFile    =   myNPM+'.zip';
                        launchPrfx      =   '';
                        launchSffx      =   '';
                        npmExtractPath  =   myNodePath+"/node_modules"
                        npmPath         =   npmExtractPath+'/npm'
                        npmTempPath     =   npmPath+'-'+myNPM.substring(1)
                        npmCommandFile  =   myNodePath+'/npm.cmd';
                        distFolder      =   'dist/'
                        break;
        default:                                                                                        //  ERROR!!!
            exit("unknown 'architecture' ... ",isOS);
    }
    
    // our App runs with its own copy of Node installed in myNodePath/node
    try         { fs.lstatSync(myNodePath); }
    catch (err) { 
        try     { fs.mkdirSync(myNodePath); }
        catch (err)    { exit("sorry can't create folder: ",myNodePath,err); }      
    }

    try{    
        fileToDown=   myNodePath+'/'+myNodeFile;
        if (!fs.existsSync(fileToDown)){
            download(nodeDistURI+myNodeFile        , fileToDown , function (err) { 
                err && me.exit(err.message,err); dwnInProg1=false; });
        } 
        else dwnInProg1=false;         
        
        fileToDown=   myNodePath+'/'+myNodeFile2;
        if (myNodeFile2 && !fs.existsSync(fileToDown)){
            download(nodeDistURI+myNodeFile2       ,myNodePath+'/'+myNodeFile2 , function (err) { 
                err && me.exit(err.message,err); dwnInProg2=false; });
        }
        else  dwnInProg2=false;
            
        fileToDown=   myNodePath+'/'+myNpnZipFile;
        if (!fs.existsSync(fileToDown)){
            download(npmDistBase+myNPM               ,myNodePath+'/'+myNpnZipFile, function (err) { 
                err && me.exit(err.message,err); dwnInProg3=false; });
        }
        else  dwnInProg3=false;
    }
    catch (err)    { exit("sorry can't download Node:",err); }
     
    

    weNeedDownload=dwnInProg1 || dwnInProg2 || dwnInProg3;
    w4it.enableAnimation();
    w4it.done(function () { return !dwnInProg1 
                                && !dwnInProg2
                                && !dwnInProg3
                                ; }
            , function (){
                var cmdLine =launchPrfx+myNodePath+'/'+myNodeFile+launchSffx
                ,   argss   = args
                ;
                argss.splice(0,1);                                                                      // remove node
                argss[0] = bootFile;                                                                    // replace myself
                
                if (weNeedDownload) _log("download finished ..");
                try {
                    if (!fs.existsSync(npmPath) && !fs.existsSync(npmTempPath)){
                        _log ("unzipping:",myNpnZipFile);
                        var child=exec("xtract",[  myNodePath+'/'+myNpnZipFile,   npmExtractPath   ],{  stdio: 'inherit' } );
                    }
                }
                catch (err){ exit("sorry can't install NPM:",err); }
        
                try{
                    if (fs.existsSync(npmTempPath) && !fs.existsSync(npmPath)){
                        _log("renaming:",npmTempPath,npmPath)
                        fs.renameSync(npmTempPath,npmPath);
                    }
                }
                catch (err){ exit("sorry can't install NPM:",err.message); }
        
                try{
                    if (!fs.existsSync(npmCommandFile)){
                        _log("coping:",npmCommandFile)
                        copyFileSync(npmPath+'/bin/npm'       ,myNodePath+'/npm');
                        copyFileSync(npmPath+'/bin/npm.cmd'   ,myNodePath+'/npm.cmd');
                    }
                }
                catch (err){ exit("sorry can't install NPM:",err.message); }
                
                if (!fs.existsSync(distFolder+myNodeVer)){
                    _log('put my own node version in list ..');
                    try {
                        linkDirSync(myNodePath,distFolder+myNodeVer);
                    }
                    catch(ex){ _err(ex.message); }    
                }
                    
                if (!fs.existsSync('node_modules')){         
                    _log("installing packages...");
                    try {
                        exec(toLocalOs(npmCommandFile),["install"],{  stdio: 'inherit' } );
                    }
                    catch(ex){ _err(ex.message); }    
                }
                
                
                var child=childProc.spawn(cmdLine,argss,{  stdio: 'inherit' } );
                child.on('error',function (err) { 
                    _err(err);    
                    me.chdir(__CWD);
                    me.exit(-123);  
                });
                child.on('exit', function (code){               
                    me.chdir(__CWD);
                    me.exit(code)   
                });
        });
    
